### PR TITLE
BlockchainObserver to WalletEventObserver refactoring

### PR DESCRIPTION
**Pull request by marekkirejczyk**
* Add TEST_KEY & TEST_KEY2 to commons

* Split into BlockchainObserver from BlockchainObserverFactory

* Add proper typing

* Fix BlockchainObserver.unsubscribe

* Rename BlockchainObserver -> WalletEventsObserver

* Simplify Subscription.remove into single unsubscribe()

* refactor WalletEventsObserverFactory integration tests

### CODE REVIEW

File: universal-login-sdk/test/helpers/createSdkWithRelayer.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+import {Wallet} from 'ethers';
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/test/helpers/createSdkWithRelayer.ts]: [Duplication](https://github.com/UniversalLogin/UniversalLoginSDK/blob/master/universal-login-sdk/test/helpers/setupSdk.ts)?

File: universal-login-sdk/test/helpers/createDeployedWallet.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import {Wallet, utils} from 'ethers';
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/test/helpers/createDeployedWallet.ts]: [Duplication](https://github.com/UniversalLogin/UniversalLoginSDK/blob/master/universal-login-sdk/test/helpers/createWallet.ts)?

File: universal-login-commons/src/core/constants/test.ts [hunk 3]
```diff
@@ -3,6 +3,10 @@ import {ETHER_NATIVE_TOKEN} from './constants';
 
 export const TEST_ACCOUNT_ADDRESS = '0x0000000000000000000000000000000000000001';
 
+export const TEST_KEY = '0x5000000000000000000000000000000000000005';
```
Comment (JustynaBroniszewska) [inline universal-login-commons/src/core/constants/test.ts]: We should stick to the names convention and name it `TEST_ACCOUNT_ADDRESS_1`, `TEST_ACCOUNT_ADDRESS_2`. 
Btw. I don't feel good with `TEST_KEY`. `TEST_PUBLIC_KEY` would be better.

File: universal-login-sdk/src/api/sdk.ts [hunk 4]
```diff
@@ -209,9 +210,9 @@ class UniversalLoginSDK {
     return this.relayerApi.cancelConnection(authorisationRequest);
   }
 
-  subscribe(eventType: string, filter: any, callback: Function) {
+  subscribe(eventType: WalletEventType, filter: WalletEventFilter, callback: WalletEventCallback) {
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/src/api/sdk.ts]: 👍

File: universal-login-sdk/src/api/sdk.ts [hunk 5]
```diff
@@ -209,9 +210,9 @@ class UniversalLoginSDK {
...
     ensure(['KeyAdded', 'KeyRemoved'].includes(eventType), InvalidEvent, eventType);
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/src/api/sdk.ts]: `ensure(eventType instanceof WalletEventType)`

File: universal-login-sdk/src/api/wallet/DeployedWallet.ts [hunk 6]
```diff
@@ -134,12 +134,12 @@ export class DeployedWallet extends AbstractWallet {
...
   }
 
-  subscribeDisconnected(onDisconnected: Function) {
-    const subscription = this.sdk.subscribe(
+  subscribeDisconnected(onDisconnected: Procedure) {
+    const unsubscribe = this.sdk.subscribe(
       'KeyRemoved',
       {contractAddress: this.contractAddress, key: this.publicKey},
-      onDisconnected,
+      () => onDisconnected(),
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/src/api/wallet/DeployedWallet.ts]: why?

File: universal-login-sdk/src/core/services/WalletService.ts [hunk 7]
```diff
@@ -149,18 +149,18 @@ export class WalletService {
...
       });
-      connectingWallet.setSubscription(subscription);
+      connectingWallet.unsubscribe = unsubscribe;
     });
   }
 
   async cancelWaitForConnection() {
     if (this.state.kind === 'Deployed') return;
-    this.getConnectingWallet().subscription && this.getConnectingWallet().subscription!.remove();
+    this.getConnectingWallet().unsubscribe && this.getConnectingWallet().unsubscribe!();
```
Comment (JustynaBroniszewska) [inline universal-login-sdk/src/core/services/WalletService.ts]: We don't need  `this.getConnectingWallet().unsubscribe && ...` `this.getConnectingWallet().unsubscribe!()` is enough.

File: universal-login-sdk/src/api/wallet/DeployedWallet.ts (hunk 6 above)
Comment (marekkirejczyk) [reply universal-login-sdk/src/api/wallet/DeployedWallet.ts]: this

File: universal-login-commons/src/core/constants/test.ts (hunk 3 above)
Comment (marekkirejczyk) [reply universal-login-commons/src/core/constants/test.ts]: Semantically that are two different things. Just for self-documenting test, we put Key, where we mean ephemeral key and account when we mean ethereum account.

File: universal-login-sdk/src/api/sdk.ts (hunk 5 above)
Comment (xitronix) [reply universal-login-sdk/src/api/sdk.ts]: Can't be use here. 
instarceof can be used with classes(on right sides) and values with any type (leftSide) of instanceof

File: universal-login-sdk/src/core/observers/WalletEventsObserverFactory.ts [hunk 8]
```diff
@@ -10,7 +11,7 @@ class WalletEventsObserverFactory extends ObserverRunner {
 
   private lastBlock: Nullable<number> = null;
 
-  constructor(private blockchainService: BlockchainService) {
+  constructor(private blockchainService: BlockchainService, private currentBlock: BlockProperty) {
```
Comment (marekkirejczyk) [inline universal-login-sdk/src/core/observers/WalletEventsObserverFactory.ts]: 👍

File: universal-login-sdk/src/core/observers/WalletEventsObserver.ts [hunk 9]
```diff
@@ -1,47 +1,43 @@
...
+import {WalletEventType, WalletEventObservableRecord} from '../models/events';
 import {parseArgs} from '../utils/events';
 import {Log} from 'ethers/providers';
 import {WalletContractInterface, BlockchainService} from '@universal-login/contracts';
 
 const eventInterface = WalletContractInterface.events;
 
 export class WalletEventsObserver {
-  private readonly callbacks: Record<WalletEventType, WalletEventCallback[]> = {
+  private readonly observableRecords: Record<WalletEventType, WalletEventObservableRecord[]> = {
```
Comment (marekkirejczyk) [inline universal-login-sdk/src/core/observers/WalletEventsObserver.ts]: observables

**Pull request merged**